src/login/pam_systemd.c

```diff
@@ -199,35 +199,56 @@ static bool display_is_local(const char *display) {
                 display[1] <= '9';
 }
 
-static int socket_from_display(const char *display, char **path) {
+static int socket_from_display(const char *display) {
+        _cleanup_free_ char *f = NULL;
         size_t k;
-        char *f, *c;
+        char *c;
+        union sockaddr_union sa;
+        socklen_t sa_len;
+        _cleanup_close_ int fd = -1;
+        int r;
 
         assert(display);
-        assert(path);
 
         if (!display_is_local(display))
                 return -EINVAL;
 
         k = strspn(display+1, "0123456789");
 
-        f = new(char, STRLEN("/tmp/.X11-unix/X") + k + 1);
+        /* Try abstract socket first. */
+        f = new(char, STRLEN("@/tmp/.X11-unix/X") + k + 1);
         if (!f)
                 return -ENOMEM;
 
-        c = stpcpy(f, "/tmp/.X11-unix/X");
+        c = stpcpy(f, "@/tmp/.X11-unix/X");
         memcpy(c, display+1, k);
         c[k] = 0;
 
-        *path = f;
+        r = sockaddr_un_set_path(&sa.un, f);
+        if (r < 0)
+                return r;
+        sa_len = r;
 
-        return 0;
+        fd = RET_NERRNO(socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC, 0));
+        if (fd < 0)
+                return fd;
+
+        r = RET_NERRNO(connect(fd, &sa.sa, sa_len));
+        if (r >= 0)
+                return TAKE_FD(fd);
+
+        if (r == -ENOENT) {
+                /* Try also non-abstract socket. */
+                memmove(sa.un.sun_path, sa.un.sun_path + 1, sa_len);
+                r = RET_NERRNO(connect(fd, &sa.sa, sa_len - 1));
```

Comment:
Looking at `sockaddr_un_set_path`, it looks like the trailing NUL is not included in the length for abstract paths, but it is included in the length for filesystem paths. I think you might need to pass `sa_len` instead of `sa_len - 1` here?

This off-by-one issue is making my head hurt.